cwl/src/main/scala/cwl/CwlCodecs.scala

```diff
@@ -1,22 +1,33 @@
 package cwl
 
+import cats.data.NonEmptyList
 import io.circe._
 import io.circe.parser._
 import io.circe.shapes._
 import io.circe.generic.auto._
 import eu.timepit.refined.string._
 import io.circe.refined._
 import io.circe.literal._
+import shapeless.Coproduct
+import cats.syntax.either._
+import cats.syntax.show._
+
 
 object CwlCodecs {
   import Implicits._
 
-  //According to automatic derivation, these instances should not be required.  But
-  //removing these breaks decodeCwl, so...
-  implicit val wfD = implicitly[Decoder[Workflow]]
-  implicit val cltD = implicitly[Decoder[CommandLineTool]]
-
-  def decodeCwl(in: String): Either[Error, Cwl] = decode[Cwl](in)
+  def decodeCwl(cwlWorkflow: String): Either[NonEmptyList[String], Cwl] = {
+    //try to parse both and combine errors if they fail
+    (decode[Workflow](cwlWorkflow), decode[CommandLineTool](cwlWorkflow)) match {
+      case (Right(wf), _) => Coproduct[Cwl](wf).asRight
```

Comment:
I was essentially suggesting this: 
```
    lazy val commandLineToolDecode = decode[CommandLineTool](cwlWorkflow) 
    (decode[Workflow](cwlWorkflow), commandLineToolDecode) match {
      case (Right(wf), _) => Coproduct[Cwl](wf).asRight
```
I *think* this works out to not evaluating the `CommandLineTool` decode if you get a `Workflow`.

Reply:
that could be instrumented with a 

```
lazy val commandLineToolDecode = {
  println("BAHH the CLT decode is being evaluated!!!")
  decode[CommandLineTool](cwlWorkflow)
}
```

Reply:
It seems the tupling evaluates each argument.

```bash
val i = 2
lazy val j = {
  println("ahh")
  3
}

(i,j) match {
  case (2,_) => println("good")
  case _ => println("bad")
}
danb@GPBD5-C11:~/cromwell$ scala r.scala 
ahh
good
```

Reply:
😢